osu.Game/Utils/ModUtils.cs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Extensions.TypeExtensions;
+using osu.Game.Rulesets.Mods;
+
+#nullable enable
+
+namespace osu.Game.Utils
+{
+    /// <summary>
+    /// A set of utilities to handle <see cref="Mod"/> combinations.
+    /// </summary>
+    public static class ModUtils
+    {
+        /// <summary>
+        /// Checks that all <see cref="Mod"/>s are compatible with each-other, and that all appear within a set of allowed types.
+        /// </summary>
+        /// <remarks>
+        /// The allowed types must contain exact <see cref="Mod"/> types for the respective <see cref="Mod"/>s to be allowed.
+        /// </remarks>
+        /// <param name="combination">The <see cref="Mod"/>s to check.</param>
+        /// <param name="allowedTypes">The set of allowed <see cref="Mod"/> types.</param>
+        /// <returns>Whether all <see cref="Mod"/>s are compatible with each-other and appear in the set of allowed types.</returns>
+        public static bool CheckCompatibleSetAndAllowed(IEnumerable<Mod> combination, IEnumerable<Type> allowedTypes)
+        {
+            // Prevent multiple-enumeration.
+            var combinationList = combination as ICollection<Mod> ?? combination.ToArray();
+            return CheckCompatibleSet(combinationList) && CheckAllowed(combinationList, allowedTypes);
+        }
+
+        /// <summary>
+        /// Checks that all <see cref="Mod"/>s in a combination are compatible with each-other.
+        /// </summary>
+        /// <param name="combination">The <see cref="Mod"/> combination to check.</param>
+        /// <returns>Whether all <see cref="Mod"/>s in the combination are compatible with each-other.</returns>
+        public static bool CheckCompatibleSet(IEnumerable<Mod> combination)
+        {
+            var incompatibleTypes = new HashSet<Type>();
```

Comment:
This hashset is added to, but never queried. (Weird that there's no inspection on this?)

In fact I'm not even sure I understand what the two hashsets are supposed to be tracking...

Reply:
I was looking at this method with the intention of combining (a portion of?) it with my follow-up function in #11642, but am also unsure how to read this code. Like I don't understand the early return or what `incomingTypes` is tracking. Is there a reason you didn't write this in a similar way to what I have in my PR?

Reply:
Re: never queried: It's supposed to be querying it in the second block, not sure how it passes the tests.

It's actually doing about the same thing your code is, just in a different way. Wouldn't mind replacing it. Though I think your code should read `type.IsInstanceOfType(m)`?

Reply:
![20210202 140924 (JetBrains Rider)](https://user-images.githubusercontent.com/191335/106555123-42634580-6560-11eb-9a20-b8fc8c750f00.png)

That's how it reads on the current version, did you mean something else?